environments/fetcher/fetcher.go

```diff
@@ -196,16 +200,50 @@ func (fetcher *Fetcher) Handler(w http.ResponseWriter, r *http.Request) {
 
 	}
 
-	// move tmp file to requested filename
-	err = os.Rename(tmpPath, filepath.Join(fetcher.sharedVolumePath, req.Filename))
+	targetFilename := filepath.Join(fetcher.sharedVolumePath, req.Filename)
+	// check file type here, if the file is a zip file unarchive it.
+	if archiver.Zip.Match(tmpPath) {
+		if req.FetchType == FETCH_SOURCE {
+			// builder accepts multiple builds at the same time, use `pkg.Name`-`pkg.ResourceVersion`
+			// to separate source pacakages for different functions.
+			targetFilename = filepath.Join(fetcher.sharedVolumePath, srcPkgFilename)
+		}
+		// unarchive tmp file to requested filename
+		err = fetcher.unarchive(tmpPath, targetFilename)
```

Comment:
Instead of unarchiving in one step, you should unarchive into a temp directory, and then rename that to the target.  The reason is that "rename" is a single atomic operation, so with this two step unarchive+rename process, we will know that if the target directory exists then it means the unarchive completed. There will never be a target directory with an incomplete set of files.

Reply:
that's a good point, I'll change it. thanks!